planner/core/rule_constant_propagation.go

```diff
@@ -0,0 +1,284 @@
+		// Attention: The function 'execOptimize' could not handle the root plan, so the parent plan could not be nil.
+		return
+	}
+	// constant propagation
+	currentPlan.constantPropagation(parentPlan, currentChildIdx, opt)
+	// recursive optimize
+	for i, children := range currentPlan.Children() {
+		cp.execOptimize(children, currentPlan, i, opt)
+	}
+}
+
+func (*constantPropagationSolver) name() string {
+	return "constant_propagation"
+}
+
+func (*baseLogicalPlan) constantPropagation(_ LogicalPlan, _ int, _ *logicalOptimizeOp) (newRoot LogicalPlan) {
+	// Only LogicalJoin can apply constant propagation
+	// Other Logical plan do nothing
+	return nil
+}
+
+// Implemented the logic of constant propagation in From List
+// Query: select * from t, (select a, b from s where s.a>1) tmp where tmp.a=t.a
+// Origin logical plan:
+/*
+            +----------------+
+            |  LogicalJoin   |
+            +-------^--------+
+                    |
+      +-------------+--------------+
+      |                            |
++-----+------+              +------+------+
+| Projection |              | TableScan   |
++-----^------+              +-------------+
+      |
+      |
++-----+------+
+| Selection  |
+|   s.a>1    |
++------------+
+*/
+//  1. 'recursiveFindAndPullUpConstantPredicates': Call this function until find selection and pull up the constant predicate layer by layer
+//     LogicalSelection: find the s.a>1
+//     LogicalProjection: get the s.a>1 and pull up it, changed to tmp.a>1
+//  2. 'addCandidateSelection': Add selection above of LogicalJoin,
+//     put all predicates pulled up from the lower layer into the current new selection.
+//     LogicalSelection: tmp.a >1
+//
+// Optimized plan:
+/*
+            +----------------+
+            |  Selection     |
+            |    tmp.a>1     |
+            +-------^--------+
+                    |
+            +-------+--------+
+            |  LogicalJoin   |
+            +-------^--------+
+                    |
+      +-------------+--------------+
+      |                            |
++-----+------+              +------+------+
+| Projection |              | TableScan   |
++-----^------+              +-------------+
+      |
+      |
++-----+------+
+| Selection  |
+|   s.a>1    |
++------------+
+*/
+// Return nil if the root of plan has not been changed
+// Return new root if the root of plan is changed to selection
+func (logicalJoin *LogicalJoin) constantPropagation(parentPlan LogicalPlan, currentChildIdx int, opt *logicalOptimizeOp) (newRoot LogicalPlan) {
+	// step1: get constant predicate from left or right according to the JoinType
+	var getConstantPredicateFromLeft bool
+	var getConstantPredicateFromRight bool
+	switch logicalJoin.JoinType {
+	case LeftOuterJoin:
+		getConstantPredicateFromLeft = true
+	case RightOuterJoin:
+		getConstantPredicateFromRight = true
+	case InnerJoin:
+		getConstantPredicateFromLeft = true
+		getConstantPredicateFromRight = true
+	default:
```

Comment:
Can this rule apply on SemiJoin?

Reply:
Yes. But should we add it now, or wait for a little bit later when there is user requirement?

Reply:
Hmm, we can add it later.

Reply:
OK